### PR TITLE
Feature/product

### PR DESCRIPTION
**Pull request by Omarmanassrah**
Implement Product:
 Is a core component of the project, designed to store and manage detailed information about products in the system. It serves 
 as  a structured way to organize data such as product names, prices, descriptions, and stock quantities. Implementing this table in 
 a database ensures that we can efficiently handle and retrieve product data through various operations.
Contain:
1)Create Product
2)Get Product  by ID
3)Update Product by ID
4)Delete Product 
5)List Products
6)Search Products

### CODE REVIEW

**Review by Mvamd12 — APPROVED** [review]

File: app/api/routes/product.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+from models import Product,ProductCreate,ProductUpdate
+
+
+
+router = APIRouter()
+    
+"""Create a Product"""
+@router.post("/products/", status_code=201)
+async def create_product(product: ProductCreate, admin: bool = Depends(get_current_admin)):
+    if not admin:
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: This one should be implemented in the get_current_admin function

File: app/api/routes/product.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        raise HTTPException(status_code=400, detail=str(e))
+    
+"""Get the Product by ID"""
+@router.get("/products/{product_id}",status_code=200 , response_model=Product)
+async def get_product_details(product_id: UUID, admin: bool = Depends(get_current_admin)):
+    try:
+        product_uuid = UUID(product_id)
+    except ValueError:
+        raise HTTPException(status_code=400,detail="Product ID must be a valid integer." )
+    product = products_db.get(product_id)
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: Move this block to be inside try

File: app/api/routes/product.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    if not product:
+        raise HTTPException(status_code=404, detail=f"Product with ID {product_id} not found.")
+    
+    return product
+
+"""Update products by the ID"""
+@router.put("/products/{product_id}", status_code=200 ,response_model=Product)
+async def update_product(product_id: UUID,update_data: ProductUpdate,admin: bool = Depends(get_current_admin)):
+    product = products_db.get(product_id)
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: Wrap this code with try except

File: app/api/routes/product.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        product.price = update_data.price
+    if update_data.stock is not None:
+        if update_data.stock < 0:
+            raise HTTPException(status_code=400,detail="Stock must be a non-negative number.")
+        product.stock = update_data.stock
+
+"""Delete product by the ID"""
+@router.delete("/products/{product_id}", status_code=200)
+async def delete_product(product_id: UUID):
+    if product_id not in products_db:
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: Add try except

File: app/api/routes/product.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            "price": product_data["price"],
+            "description": product_data.get("description", ""), 
+            "stock": product_data.get("stock", 0),  
+            "isAvailable": product_data.get("isAvailable", True),  
+            "created_at": product_data["created_at"]
+        }
+        products_list.append(product_dict)
+    return products_list
+
+class ProductSearchParams(BaseModel):
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: Move it to models file

**Review by FaresAbuIram — CHANGES_REQUESTED** [review]

File: app/api/routes/product.py (hunk 1 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 2 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 3 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 4 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 5 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    try:
+        product = products_db.get(product_id)
+        raise HTTPException(status_code=404,detail=f"Product with ID {product_id} not found.")
+        if update_data.name:
+            if any(existing_product['name'].lower() == update_data.name.lower() for existing_product in products_db.values()):
+                raise HTTPException(status_code=400,detail=f"Product name '{update_data.name}' already exists. Please use a unique name.")
+        product.name = update_data.name
+        if update_data.price is not None:
+            if update_data.price < 0:
+                raise HTTPException(status_code=400,detail="Price must be a non-negative number.")
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: This exception is already exist on update product model, remove it

File: app/api/routes/product.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        product.price = update_data.price
+        if update_data.stock is not None:
+            if update_data.stock < 0:
+                raise HTTPException(status_code=400,detail="Stock must be a non-negative number.")
```
Comment (FaresAbuIram) [inline app/api/routes/product.py]: The same

File: app/api/routes/product.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            "description": product_data.get("description", ""), 
+            "stock": product_data.get("stock", 0),  
+            "isAvailable": product_data.get("isAvailable", True),  
+            "created_at": product_data["created_at"]
+        }
+        products_list.append(product_dict)
+    return products_list
+    
+""" Search for Product"""
+@router.get("/products/search",status_code=200,response_model=dict)
```
Comment (FaresAbuIram) [inline app/api/routes/product.py:98]: This logic will be changed when connecting DB, i will ignore it for now

**Review by FaresAbuIram — CHANGES_REQUESTED** [review]

File: app/api/routes/product.py (hunk 6 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 7 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py]: Done

File: app/api/routes/product.py (hunk 8 above)
Comment (Omarmanassrah) [reply app/api/routes/product.py:98]: Forsure

**Review by FaresAbuIram — APPROVED** [review]

**Pull request merged**